File: TinkoffChat/Themes/ThemesViewController.m
//
//  ThemesViewController.m
//  TinkoffChat
//
//  Created by Олег Самойлов on 17/03/2018.
//  Copyright © 2018 Oleg Samoylov. All rights reserved.
//

#import "Themes.h"
#import "ThemesViewController.h"
#import "ThemesViewControllerDelegate.h"

@implementation ThemesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _model = [[Themes alloc]initWithTheme1:[UIColor colorWithRed:(121.0/255.0) green:(214.0/255.0) blue:(249.0/255.0) alpha:1.0]
                                    theme2:[UIColor colorWithRed:(184.0/255.0) green:(226.0/255.0) blue:(151.0/255.0) alpha:1.0]
                                    theme3:[UIColor colorWithRed:1.0 green:(244.0/255.0) blue:(100.0/255.0) alpha:1.0]];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INTERACTIVE, 0), ^{
        NSData *colorData = [[NSUserDefaults standardUserDefaults] objectForKey:@"themeColor"];
        if (colorData != nil) {
            UIColor *color = [NSKeyedUnarchiver unarchiveObjectWithData:colorData];
            dispatch_async(dispatch_get_main_queue(), ^{
                self.view.backgroundColor = color;
            });
        }
    });
    
    NSData *colorData = [[NSUserDefaults standardUserDefaults] objectForKey:@"themeColor"];
    if (colorData != nil) {
        UIColor *color = [NSKeyedUnarchiver unarchiveObjectWithData:colorData];
        self.view.backgroundColor = color;
    }
}

- (void)dealloc {
    [_model release];
    
    [super dealloc];
}

- (Themes *)model {
    return _model;
}

- (void)setModel:(Themes *)model {
    if (_model != model) {
        [_model release];
        _model = [model retain];
    }
}

- (id<ThemesViewControllerDelegate>)delegate {
    return _delegate;
}

- (void)setDelegate:(id<ThemesViewControllerDelegate>)delegate {
    if (_delegate != delegate)
        _delegate = delegate;
}

- (IBAction)didCloseBarButtonItemTap {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)didThemeButtonTap:(UIButton *)sender {
    if (_delegate == nil)
        return;
    
    switch (sender.tag) {
        case 1:
            [_delegate themesViewController:self
                             didSelectTheme:self.model.theme1];
            break;
        case 2:
            [_delegate themesViewController:self
                             didSelectTheme:self.model.theme2];
            break;
        case 3:
            [_delegate themesViewController:self
                             didSelectTheme:self.model.theme3];
            break;
        default:
            break;
    }
}

@end

File: TinkoffChat/PresentationLayer/Modules/Themes/ThemesViewController.swift
//
//  ThemesViewController.swift
//  TinkoffChat
//
//  Created by Олег Самойлов on 21/03/2018.
//  Copyright © 2018 Oleg Samoylov. All rights reserved.
//

import UIKit

protocol ThemesSelector: class {
    func didThemeButtonTap(_ sender: UIButton)
}

class ThemesViewController: UIViewController, ThemesSelector {
    
    // MARK: - Dependencies
    
    private let model: IThemesModel
    var mock: ThemesSelector?
    
    // MARK: - Initializers
    
    init(model: IThemesModel) {
        self.model = model
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: - Outlet
    
    @IBOutlet var buttonsWithTag: [UIButton]!
    
    // MARK: - UIViewController
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.navigationBar.prefersLargeTitles = true
        
        configureNavigationPane()
        
        DispatchQueue.global(qos: .userInteractive).async {
            if let theme = UserDefaults.standard.colorForKey(key: "themeColor") {
                DispatchQueue.main.async {
                    self.view.backgroundColor = theme
                }
            }
        }
    }
    
    // MARK: - Action
    
    @IBAction func didThemeButtonTap(_ sender: UIButton) {
        switch sender.tag {
        case 1:
            model.closure(self, model.theme1)
        case 2:
            model.closure(self, model.theme2)
        case 3:
            model.closure(self, model.theme3)
        default:
            break
        }
        
        mock?.didThemeButtonTap(sender)
    }
    
    @IBAction func didInvertButtonTap() {
        guard let color = view.backgroundColor else { return }
        
        model.invert(color)
        model.closure(self, color.invertedColor)
    }
    
    // MARK: - Private methods
    
    private func configureNavigationPane() {
        navigationItem.title = "Темы"
        
        let rightItem = UIBarButtonItem(title: "Закрыть",
                                        style: .plain,
                                        target: self,
                                        action: #selector(close))
        navigationItem.setRightBarButton(rightItem, animated: true)
    }
    
    @objc private func close() {
        dismiss(animated: true)
    }
    
    // MARK: - Deinitializers
    
    deinit {
        buttonsWithTag = nil
    }
}
